sdks/java/extensions/zetasketch/src/test/java/org/apache/beam/sdk/extensions/zetasketch/HllCountTest.java

```diff
@@ -0,0 +1,373 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.zetasketch;
+
+import com.google.zetasketch.HyperLogLogPlusPlus;
+import com.google.zetasketch.shaded.com.google.protobuf.ByteString;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import org.apache.beam.sdk.Pipeline.PipelineExecutionException;
+import org.apache.beam.sdk.testing.NeedsRunner;
+import org.apache.beam.sdk.testing.PAssert;
+import org.apache.beam.sdk.testing.TestPipeline;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.TypeDescriptor;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Tests for {@link HllCount}. */
+@RunWith(JUnit4.class)
+public class HllCountTest {
```

Comment:
The tests currently included in `HllCountTest` are intended to be at unit-test level. I looked into the config files you linked here and it seems be for Dataflow ValidateRunner tests. I am not sure if we want to add these tests into that list, because
- `HllCount` itself lives in a extension module that is outside of `:sdks:java:core`. If we want to run them on Dataflow post-commit tests we will need the `:runners:google-cloud-dataflow-java` module to depend on this extension module, which seems weird to me.
- `HllCount` is just another aggregation transform built on top on the `Combine` transform, similar to `Sum`, `Mean`, or `Latest` (a more comprehensive list: https://beam.apache.org/documentation/transforms/java/overview/#aggregation). I checked the test list here https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/3162/testReport/org.apache.beam.sdk.transforms/ and non of those transforms are included (only fundamental transforms like `ParDo`, `Combine`, `GroupByKey` etc. are tested).

However, I totally agree that we should have some tests against the target DataflowRunner for this feature. I am working on an integration test that tests not only `HllCount` works on Dataflow, but also it can interoperate with `BigQuery` with the same sketch format. There are some work need to be done, and I would like to put that integration test in a separate PR, to avoid blocking the review progress of this PR.

Reply:
> The tests currently included in HllCountTest are intended to be at unit-test level.

All test cases in HllCountTest are marked as `NeedsRunner`, which means these tests are required to run with a runner. Thus, I don't think these tests are at UT level. And if there is no test target running these tests, I'm not sure what the purpose of tests are.

How to construct tests is depended on how you scope your feature. If you think your feature is dataflow only, then you should run your tests with dataflow runner. You can try to define a test target in your module's gradle file and make it depend on dataflow-runner project, but I'm not sure whether it will introduce circular dependencies.

As I mentioned above, I totally agree it's better for us to have ITs with BQ if BQ is the major use case. But I would prefer having tests inside this PR. Otherwise, seems like we will not have any proof saying that this transform works, right?

Reply:
> All test cases in HllCountTest are marked as NeedsRunner, which means these tests are required to run with a runner. Thus, I don't think these tests are at UT level.

Well, I agree with you that maybe we should not call these tests unit tests because they are not testing individual components and requires a runner. But according to the [Beam Contribution Testing Guide](https://cwiki.apache.org/confluence/display/BEAM/Contribution+Testing+Guide#ContributionTestingGuide-Unit), if we categorize all our tests into 3 categories, meaning unit test, ValidatesRunner test, and IT, NeedsRunner test still falls into unit test... Anyway, I will just call these tests NeedsRunner test from now on.

> And if there is no test target running these tests, I'm not sure what the purpose of tests are.

Actually NeedRunner tests are included in our Java Precommit tests on Jenkins (see the [test for this PR](https://builds.apache.org/job/beam_PreCommit_Java_Commit/7044/testReport/)), using DirectRunner, I think. I believe the test target running these tests is `:runners:direct-java:needsRunner`.

> As I mentioned above, I totally agree it's better for us to have ITs with BQ if BQ is the major use case. But I would prefer having tests inside this PR. Otherwise, seems like we will not have any proof saying that this transform works, right?

Agree. Currently there is no proof that this feature is working on DataflowRunner in the form of a test, which is not good. I will write the ITs and include them along with this PR.

> How to construct tests is depended on how you scope your feature. If you think your feature is dataflow only, then you should run your tests with dataflow runner. You can try to define a test target in your module's gradle file and make it depend on dataflow-runner project, but I'm not sure whether it will introduce circular dependencies.

This is how I want to add the ITs. I will define the ITs in this module instead of in `:runners:google-cloud-dataflow-java`, and add a dependency to DataflowRunner as a test dependency.

Reply:
Hi. The integration tests (running on Dataflow and BigQuery) in added to this PR! PTAL.

Here is the link showing the tests run successfully on Jenkins as a Java PostCommit test: https://builds.apache.org/job/beam_PostCommit_Java_PR/221/testReport/org.apache.beam.sdk.extensions.zetasketch/

Reply:
I'll resolve this thread since the tests have been added. If you have any comment on the integration tests implementation, please feel free to start a new thread to discuss that!